File: d01/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

func parseFileToSlice(file *os.File) []int64 {
	var results []int64
	var temp int64

	sc := bufio.NewScanner(file)

	// Check each lines and add numbers to temp
	// If line is empty, reset temp and append result to slice
	for sc.Scan() {
		if line := sc.Text(); line != "" {
			if n, err := strconv.ParseInt(line, 10, 64); err == nil {
				temp += n
			}
		} else {
			results = append(results, temp)
			temp = 0
		}
	}

	return results
}

func findMaximum(arr []int64) (int64, int) {
	var max int64 = 0
	var index int = 0

	max = arr[0]

	// Find the largest numbers in slice
	for i := range arr {
		if arr[i] > max {
			max = arr[i]
			index = i
		}
	}

	return max, index
}

func first(file *os.File) int64 {
	// Parse file to slice
	lines := parseFileToSlice(file)
	max, _ := findMaximum(lines)
	return max
}

func second(file *os.File) int64 {
	lines := parseFileToSlice(file)
	var n int = 3
	topThree := [3]int64{}
	var result int64 = 0

	for i := 0; i < n; i += 1 {
		max, index := findMaximum(lines)

		log.Println(max)

		topThree[i] = max
		next := index + 1
		lines = append(lines[:index], lines[next:]...)
		result += max
	}

	return result
}

func main() {
	file, err := os.Open("../input.txt")
	if err != nil {
		panic(err)
	}

	defer file.Close()

	// fmt.Println("The maxium calories:", first(file))
	fmt.Println("The sum of top three calories:", second(file))
}

File: d01/rs/src/main.rs
use std::fs;

// fn first_part_a(content: &String) -> i32 {
//     let parts = content.split("\n\n");
//     let mut results: Vec<i32> = vec![];
//     let mut max: i32 = 0;
//
//     for part in parts {
//         let mut sum: i32 = 0;
//         let lines = part.split("\n");
//
//         for line in lines {
//             if line != "" {
//                 let num: i32 = line.parse().expect("should able to parse string to i32");
//                 sum += num
//             }
//         }
//
//         results.push(sum)
//     }
//
//     for i in results {
//         if i > max {
//             max = i
//         }
//     }
//
//     return max;
// }

fn first_part_b(content: &str) -> i32 {
    content
        .split("\n\n")
        .map(|part| {
            part.split("\n")
                .filter(|line| !line.is_empty())
                .map(|n| {
                    n.parse::<i32>()
                        .expect("should able to parse string to i32")
                })
                .sum()
        })
        .max()
        .unwrap_or_default()
}

// fn second_part_a(content: &str) -> i32 {
//     let mut results = vec![];
//     let mut top_three: [i32; 3] = [0, 0, 0];
//
//     for part in content.split("\n\n") {
//         let value: i32 = part
//             .lines()
//             .filter(|line| !line.is_empty())
//             .map(|n| n.parse::<i32>().unwrap())
//             .sum::<i32>();
//
//         results.push(value);
//     }
//
//     results.sort_by(|a, b| b.cmp(a));
//
//     for i in 0..3 {
//         top_three[i] = results[i]
//     }
//
//     return top_three.iter().sum();
// }

fn second_part_b(content: &str) -> i32 {
    let mut top_three: [i32; 3] = [0, 0, 0];

    for part in content.split("\n\n") {
        let value: i32 = part
            .lines()
            .filter(|line| !line.is_empty())
            .map(|n| n.parse::<i32>().unwrap())
            .sum::<i32>();

        for i in 0..3 {
            if value > top_three[i] {
                top_three.swap(i, 2);
                top_three.swap(1, 2);
                top_three.swap(0, 1);
                top_three[i] = value;
                break;
            }
        }
    }

    return top_three.iter().sum();
}

fn main() {
    let content = fs::read_to_string("../input.txt").expect("should able to read the file");
    let n = first_part_b(&content);
    let m = second_part_b(&content);
    println!("The maxium calories is {}", n);
    println!("The sum of top three calories is {}", m);
}
